lib/Optimizer/Transforms/ConstPropComplex.cpp

```diff
@@ -0,0 +1,200 @@
+/*******************************************************************************
+ * Copyright (c) 2022 - 2024 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ ******************************************************************************/
+
+#include "PassDetails.h"
+#include "cudaq/Optimizer/Builder/Intrinsics.h"
+#include "cudaq/Optimizer/Dialect/CC/CCOps.h"
+#include "cudaq/Optimizer/Dialect/Quake/QuakeOps.h"
+#include "cudaq/Optimizer/Transforms/Passes.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Dominance.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+#include "mlir/Transforms/Passes.h"
+
+namespace cudaq::opt {
+#define GEN_PASS_DEF_CONSTPROPCOMPLEX
+#include "cudaq/Optimizer/Transforms/Passes.h.inc"
+} // namespace cudaq::opt
+
+#define DEBUG_TYPE "const-prop-complex"
+
+using namespace mlir;
+
+namespace {
+
+// Fold complex.create ops if the arguments are constants.
+class ComplexCreatePattern : public OpRewritePattern<complex::CreateOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(complex::CreateOp create,
+                                PatternRewriter &rewriter) const override {
+    auto re = create.getReal();
+    auto im = create.getImaginary();
+    auto reCon = re.getDefiningOp<arith::ConstantOp>();
+    auto imCon = im.getDefiningOp<arith::ConstantOp>();
+    if (reCon && imCon) {
+      auto aa = ArrayAttr::get(
+          rewriter.getContext(),
+          ArrayRef<Attribute>{reCon.getValue(), imCon.getValue()});
+      rewriter.replaceOpWithNewOp<complex::ConstantOp>(create, create.getType(),
+                                                       aa);
+      return success();
+    }
+    return failure();
+  }
+};
+
+// Fold floating point cast ops if the argument is constant.
+class FloatCastPattern : public OpRewritePattern<cudaq::cc::CastOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(cudaq::cc::CastOp cast,
+                                PatternRewriter &rewriter) const override {
+    auto val = cast.getOperand();
+    auto valCon = val.getDefiningOp<arith::ConstantFloatOp>();
+    if (valCon) {
+      auto fTy = dyn_cast<FloatType>(cast.getType());
+      if (fTy == rewriter.getF64Type()) {
+        auto v = valCon.value().convertToFloat();
```

Comment:
Do we have code coverage on this line? I seem to recall some issue with this call. Unfortunately, I don't remember the specifics.

Reply:
I have tests in quake for this, AFAIK we don't generate this in code yet. But it will help after https://github.com/NVIDIA/cuda-quantum/issues/1955 is addressed.

But I think I am missing a safety check for the operand's type here. 

Added another test (I only had a test for one direction of the cast before). I remember seeing problems with those functions too - I think `valCon.value().convertToFloat()` fails if the value is double, I added an extra check to ensure that it is indeed a float.